MatrixFactorizations Package

kellerlv opened the pull request:
A new matrix factorizations package, a package for computing with length d matrix factorizations. Authors David Favero, Sasha Pevzner, Tim Tribone, and Keller VandeBogert.

kellerlv committed:
MatrixFactorization Package

Package and documentation files added

kellerlv committed:
Merge branch 'development' into MatrixFactorizations

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations.m2:
```diff
@@ -0,0 +1,136 @@
+            Email => "kvandebo@nd.edu", 
+            HomePage => "https://sites.google.com/view/kellervandebogert/home"}
+    },
+    Headline => "a package for computing with length d matrix factorizations",
```
"a package for" here is redundant (and will cause one of the Core tests to fail).  Also, what's d?  Maybe "matrix factorizations of a particular length" or something like that?

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations.m2:
```diff
@@ -0,0 +1,136 @@
+	"CompleteIntersectionResolutions",
+	"TensorComplexes"
+    },
+    DebuggingMode => true
```
`DebuggingMode` should be false, or even better, removed entirely since false is the default

d-torrance replied:
`newPackage` should also have a `Keywords` option.  The value is a list of strings, ideally a small subset of the existing keywords, which are the headings on the [packages provided with Macaulay2](https://macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_packages_spprovided_spwith_sp__Macaulay2.html) documentation page

kellerlv replied:
Sorry about that. I've added those things now and fixed the typos.

d-torrance reviewed (changes requested):
Thanks for the contribution!

Without diving into the code too much yet, here are a couple quick comments.  Also, there are a couple typos:
```
M2/Macaulay2/packages/MatrixFactorizations/KellerTests.m2:671: documnetation ==> documentation
M2/Macaulay2/packages/MatrixFactorizations/MatrixFactorizationsDOC.m2:707: followig ==> following
```

kellerlv committed:
Typos fixes, updated package headlines

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations.m2:
```diff
@@ -0,0 +1,136 @@
+        "Complexes",
+	"CompleteIntersectionResolutions",
+	"TensorComplexes"
+    }
```
There's no reason to duplicate the packages in `PackageExports` and `PackageImports`.  Do you want to export those packages' symbols to the user, or just use them internally?

kellerlv replied:
When this was initially started, the "Complexes" package was not the default. So it's probably unnecessary to load that now. And yes the other two packages are only called internally -- I was loading them by default for my own testing purposes, but I don't think it's necessary to export their symbols automatically for users.

mahrud replied:
It is still necessary to load or export Complexes if the package uses it.

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations/CIResCompatibility.m2:
```diff
@@ -0,0 +1,240 @@
+needsPackage "CompleteIntersectionResolutions"
+needsPackage "PushForward"
```
`CompleteIntersectionResolutions` will already be pulled in from `PackageImports` or `PackageExports`.  I'd also move `PushForward` to one of those.

kellerlv replied:
Yes PushForward should be moved to just the package imports again.

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations/CIResCompatibility.m2:
```diff
@@ -0,0 +1,240 @@
+    Hn := new HashTable from for i from 0 to length Ln-1 list Ln_i => H#((keys H)#i);
+    (lo,hi) := concentration C;
+    Lo := select(toList(lo-1..hi+1),i->odd(i));
+    Le := select(toList(lo-1..hi+1),i->even(i));
```
Minor quibble: `i -> odd i` and `i -> even i` can be replaced with just `odd` and `even`, resp.

kellerlv replied:
Oh yeah I'm not sure why I wrote it like that.

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations/KellerTests.m2:
```diff
@@ -0,0 +1,964 @@
+
+TEST ///
+--playing around with adjoining roots of unity
+///
```
Were there intended to be tests here?

kellerlv replied:
I have a memory of writing a bunch of tests for this. I need to look into if I accidentally pushed an older version of the testing file.

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations/MF_functions.m2:
```diff
@@ -0,0 +1,1152 @@
+--take this one out when putting above back in:
+--needs "ZZdFactorizations.m2"
+needsPackage "TensorComplexes" --Koszul factorization method needs this
+needsPackage "Cyclotomic" --for adjoining roots of unity
```
These should also be moved to `PackageImports` or `PackageExports`.

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations/MF_functions.m2:
```diff
@@ -0,0 +1,1152 @@
+    )
+
+--euler characteristic of a ZZdfactorization. Caveart: in order to be well-defined, the input should be a complex
+eulerChi = method();
```
`euler` already exists in Core for the Euler characteristic of various classes.  You could overload that instead of defining a new method function

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations/ZZdFactorizations.m2:
```diff
@@ -0,0 +1,1754 @@
+needsPackage "Complexes"
```
This also should be in `PackageImports` or `PackageExports`

d-torrance commented on M2/Macaulay2/packages/MatrixFactorizations/ZZdFactorizations.m2:
```diff
@@ -0,0 +1,1754 @@
+net ZZdFactorization := C -> (
+     (lo,hi) := (0,C.period);
+     if lo > hi then 
+         error "In a complex, lo <= hi should always hold in the concentration"
```
Error messages usually begin with a lowercase letter

kellerlv replied:
Actually that's a typo on my part anyway, it should say "factorization" not "complex". I'm going to wait and see if this final set of checks goes through successfully and then I'll implement these changes along with fixing any errors that might pop up if it fails.

d-torrance reviewed (changes requested):
Thanks for the quick response to the previous points!  I dug through the code a bit more and have a few more suggestions

kellerlv committed:
Made suggested changes

moved more things to packageImports, streamlined some spots, overloaded euler, added tests for adjoinRoot and pruning

d-torrance reviewed (approved):
Thanks!
